### ISSUE TITLE
Cannot configure Detekt task options individually

### ISSUE THREAD

**Issue by damianw** [detekt/detekt#3866]
From what I can tell, it's not currently possible to configure some options on individual `Detekt` tasks, for example:

```kotlin
tasks.withType<Detekt>().configureEach {
    val myCustomDestinationForThisTaskXmlReport = somethingBasedOnTheTask(this)
    val myCustomDestinationForThisTaskCustomReport = somethingElseBasedOnTheTask(this)
    reports {
        xml {
            destination = myCustomDestinationForThisTaskXmlReport
        }
        custom {
            reportId = "MyCustomReport"
            destination = myCustomDestinationForThisTaskCustomReport
        }
        // Other possible customizations
    }
    // Other custom options
}
```

The use case here is specifying custom variant-specific destinations for standard and custom reports.

## Expected Behavior
Configuring tasks in a manner such as the example should configure those tasks as expected.

## Observed Behavior
There seem to be two issues:
1. The configuration order of `Detekt` tasks puts the [Detekt internal configuration block](https://github.com/detekt/detekt/blob/4d4a11a054c35b1f077ef6c326a63e9bbdfc270d/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/SharedTasks.kt#L11-L28) _last_, so it overrides the configuration from `configureEach`. (This was actually surprising to me - I had assumed that the block passed to `register` would be executed before other configuration actions... so TIL I guess?)
2. The [assignment of](https://github.com/detekt/detekt/blob/4d4a11a054c35b1f077ef6c326a63e9bbdfc270d/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektAndroid.kt#L119) `reports = extension.reports` re-uses the same `DetektReports` object, thus causing it to be shared across all tasks. Configuring one task will result in them all being configured with the same reports.

## Steps to Reproduce

Configure Detekt tasks in a project with different outputs - for example, based on the task name:

```kotlin
tasks.withType<Detekt>().configureEach {
    reports {
        xml {
			enabled = true
            destination = File(project.buildDir, "$name/detekt.xml")
        }
    }
}
```

Observe that all of the tasks will output to the same location.

## Context
I am trying to customize output location of the report files - specifically of custom reports, in order to include the variant name in the path. Before type resolution was added officially, we (on an Android project) had a variant-based output directory structure for all of the reports, `build/reports/detekt/<variantName>/<reportName>`, (including for custom reports) which I am trying to preserve. I have found that I'm able to work around the issue with a separate "configuration task" that I can depend the `Detekt` tasks on, in order to set a new `DetektReports` instance before the `Detekt` task runs. 🙃

## Your Environment
* Version of detekt used: `1.17.1`
* Version of Gradle used (if applicable): `6.9`
* Operating System and version: Mac OS 11.4

**Comment by cortinico** [comment]
Thanks for the report, really appreciate 🙏 

>     2. The [assignment of](https://github.com/detekt/detekt/blob/4d4a11a054c35b1f077ef6c326a63e9bbdfc270d/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektAndroid.kt#L119) `reports = extension.reports` re-uses the same `DetektReports` object, thus causing it to be shared across all tasks. Configuring one task will result in them all being configured with the same reports.

You're right, that's not ideal. 
This is actually addressed by #3687 (ping @3flex for moving that forward).

>     1. The configuration order of `Detekt` tasks puts the [Detekt internal configuration block](https://github.com/detekt/detekt/blob/4d4a11a054c35b1f077ef6c326a63e9bbdfc270d/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/SharedTasks.kt#L11-L28) _last_, so it overrides the configuration from `configureEach`. (This was actually surprising to me - I had assumed that the block passed to `register` would be executed before other configuration actions... so TIL I guess?)
 
I'm unsure about how to address this. I'm also really surprised that this is the case. 
I believe that #3687 should unblock you on this front as well, but I'd like to understand if we have an underlying issue with our Gradle setup or not 🤔

**Comment by 3flex** [comment]
This should be re-verified on 1.19.0-RC1, if the issue remains please advise and I'll investigate.

**Comment by BraisGabin** [comment]
@damianw Could you check if you have the same issue with detekt 1.19.0?

**Comment by cortinico** [comment]
Closing as this was fixed in 1.19.0
